### PR TITLE
Support compare diff URLs in `prevent-pr-commit-link-loss`

### LINKED ISSUE #3528: Support compare diff URLs in `prevent-pr-commit-link-loss`
Copy-paste this URL in a comment a click "Preview"

```
https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191
```

GitHub changes the URL to the following, breaking it

```
https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0diff-6be2971b2bb8dbf48d15ff680dd898b0R191
```

This is a similar GitHub bug to https://github.com/sindresorhus/refined-github/issues/2327, which was fixed by https://github.com/sindresorhus/refined-github/pull/3085 ([prevent-pr-commit-link-loss](https://github.com/sindresorhus/refined-github/blob/master/source/features/prevent-pr-commit-link-loss.tsx)), so it'd make sense to extend that feature

### PR DESCRIPTION
**Pull request by esakki-selvaraj**
Thanks for contributing! 🍄

1. LINKED ISSUES:
   Closes #3528 

2. TEST URLS:

Copy-paste these into a comment box
```
Non-buggy URL
https://github.com/sindresorhus/refined-github/compare/20.9.17...master

One line selected:
https://github.com/sindresorhus/refined-github/compare/20.9.11...master#diff-082c28d748ad2e3eecc5508d740d9417R29

Multiple lines selected
https://github.com/sindresorhus/refined-github/compare/template...835b0a0d6d1b5cbb60e7a93cf43b3535602e599e#diff-6bbbbaf49258c5f6252cb684d47d16f4R2-R3
```

3. SCREENSHOT:

### CODE REVIEW

File: source/github-helpers/index.ts [hunk 1]
```diff
@@ -98,6 +98,9 @@ const escapeRegex = (string: string): string => string.replace(/[\\^$.*+?()[\]{}
 const prCommitPathnameRegex = /[/][^/]+[/][^/]+[/]pull[/](\d+)[/]commits[/]([\da-f]{7})[\da-f]{33}(?:#[\w-]+)?\b/; // eslint-disable-line unicorn/better-regex
 export const prCommitUrlRegex = new RegExp('\\b' + escapeRegex(location.origin) + prCommitPathnameRegex.source, 'gi');
 
+const prComparePathnameRegex = /[/][^/]+[/][^/]+[/]compare[/]([v\da-f.]{7})...(master|[v\da-f.]{7})#?diff-([\da-f]{7})[\da-f]{25}(?:[#\d\w-]+)?\b/; // eslint-disable-line unicorn/better-regex
```
Comment (fregante) [inline source/github-helpers/index.ts:101]: The regex is incorrect. You can compare any branches and commits, not just `v*` tags and `master`

Probably a generic `(.+)` is enough to capture the entire comparison
```diff
- ([v\da-f.]{7})...(master|[v\da-f.]{7})
+ (.+)
```

The `#` **must** be there, so

```diff
- #?diff
+ #diff
```

I think the `(?:[#\d\w-]+)?` part refers to the URL hash, which in this case is already `#diff-*` so it should be dropped

File: source/github-helpers/index.ts [hunk 2]
```diff
@@ -106,6 +109,14 @@ export function preventPrCommitLinkLoss(url: string, pr: string, commit: string,
 
 	return `[\`${commit}\` (#${pr})](${url})`;
 }
+//To be  used as replacer callback in string.replace() for compare links
```
Comment (fregante) [inline source/github-helpers/index.ts:112]: Space above
```suggestion

//To be  used as replacer callback in string.replace() for compare links
```

File: source/github-helpers/index.ts [hunk 3]
```diff
@@ -106,6 +109,14 @@ export function preventPrCommitLinkLoss(url: string, pr: string, commit: string,
...
+export function preventPrCompareLinkLoss(url: string, compareFile1: string, compareFile2: string, compare: string, index: number, fullText: string): string {
+	if (fullText[index + url.length] === ')') {
+		return url;
+	}
+
+	return `[\`${compare}\` (${compareFile1}-${compareFile2})](${url})`;
```
Comment (fregante) [inline source/github-helpers/index.ts:118]: We can capture the whole `#diff` part and just append it. This is not a commit hash and it's meaningless.
```suggestion
	return `[\`${compare}\`${diff}](${url})`;
```

It would appear as: [`a...b`#diff-erxvgbhjnm](#)

File: source/features/prevent-pr-commit-link-loss.tsx [hunk 4]
```diff
@@ -7,32 +7,35 @@ import * as pageDetect from 'github-url-detection';
...
 			<button type="button" className="btn btn-sm primary flash-action rgh-fix-pr-commit-links">Fix link</button>
 		</div>
 	);
 }
 
-const updateUI = debounceFn(({delegateTarget: field}: delegate.Event<InputEvent, HTMLTextAreaElement>): void => {
+const updateUI = debounceFn(({ delegateTarget: field }: delegate.Event<InputEvent, HTMLTextAreaElement>): void => {
 	// The replacement logic is not just in the regex, so it alone can't be used to detect the need for the replacement
-	if (field.value === field.value.replace(prCommitUrlRegex, preventPrCommitLinkLoss)) {
+	if (field.value === field.value.replace(prCommitUrlRegex, preventPrCommitLinkLoss) && field.value === field.value.replace(prCompareUrlRegex, preventPrCompareLinkLoss)) {
```
Comment (fregante) [inline source/features/prevent-pr-commit-link-loss.tsx:30]:
```suggestion
	if (field.value === field.value.replace(prCommitUrlRegex, preventPrCommitLinkLoss).replace(prCompareUrlRegex, preventPrCompareLinkLoss)) {
```

File: source/features/prevent-pr-commit-link-loss.tsx [hunk 5]
```diff
@@ -7,32 +7,35 @@ import * as pageDetect from 'github-url-detection';
...
 		getUI(field).remove();
-	} else if (pageDetect.isNewIssue() || pageDetect.isCompare()) {
+	}
+	else if (pageDetect.isNewIssue() || pageDetect.isCompare()) {
 		select('file-attachment', field.form!)!.append(
 			<div className="m-2">{getUI(field)}</div>
 		);
-	} else {
+	}
+	else {
```
Comment (fregante) [inline source/features/prevent-pr-commit-link-loss.tsx:38]: Remember to disable auto-formatting in your editor. All of these changes are unrelated.

File: source/github-helpers/index.ts (hunk 1 above)
Comment (esakki-selvaraj) [reply source/github-helpers/index.ts:101]: https://github.com/sindresorhus/refined-github/compare/20.9.11...masterdiff-67ba01fa3e792891b585fa43e92c78ea

This link doesn't have a hash

File: test/helpers.ts [hunk 6]
```diff
@@ -208,4 +208,28 @@ test('preventPrCommitLinkLoss', t => {
 		'lorem ipsum dolor [`1da152b` (#3205)](https://github.com/sindresorhus/refined-github/pull/3205/commits/1da152b3f8c51dd72d8ae6ad9cc96e0c2d8716f5#diff-932095cc3c0dff00495b4c392d78f0afR60) some random string',
 		'It should not apply it twice even with hashes'
 	);
+
+	t.is(
+		replace('https://github.com/sindresorhus/refined-github/commit/cb44a4eb8cd5c66def3dc26dca0f386645fa29bb'),
+		'https://github.com/sindresorhus/refined-github/commit/cb44a4eb8cd5c66def3dc26dca0f386645fa29bb',
+		'It should not affect non PR commit URLs'
```
Comment (fregante) [inline test/helpers.ts:215]:
```suggestion
		replace('https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0'),
		'https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0',
		'It should not affect compare URLs without a diff hash'
```

File: source/github-helpers/index.ts (hunk 1 above)
Comment (fregante) [reply source/github-helpers/index.ts:101]: Re-read the issue that you’re trying to fix. That link does not have the hash… because it’s broken. This feature is meant to avoid such broken link.

Comment (fregante) [reply source/github-helpers/index.ts:101]: Even the “test URLs” you provided don’t make sense here. This feature acts in the comment box, not on those URLs. Refer to https://github.com/sindresorhus/refined-github/pull/3223 to see how this feature should be tested.

File: source.pem [hunk 7]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (notlmn) [inline source.pem:1]: __Caution__: _DO NOT_ commit keys on GitHub. Remove this from here, and make sure you invalidate this public-private key pair as soon as possible.

**Review by notlmn — CHANGES_REQUESTED** [review]

File: source.pem (hunk 7 above)
Comment (esakki-selvaraj) [reply source.pem:1]: @notlmn how to do that??

File: source/features/batch-mark-files-as-viewed.tsx [hunk 8]
```diff
@@ -26,10 +26,10 @@ function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
 	const previousFileState = isChecked(previousFile);
 	const thisFile = event.delegateTarget.closest<HTMLElement>('.js-file')!;
 	const files = select.all('.js-file');
-	const selectedFiles = files.slice(
+	const selectedFiles = files.slice(...[
 		files.indexOf(previousFile) + 1,
 		files.indexOf(thisFile) + 1
-	);
+	].sort((a, b) => a - b));
```
Comment (fregante) [inline source/features/batch-mark-files-as-viewed.tsx:32]: I suggest you use GitHub Desktop or similar to commit only the files and lines that are necessary — and then look at the PR and see if there's any unwanted code. We shouldn't have to comment on this. This file is unrelated

File: source/github-helpers/index.ts [hunk 9]
```diff
@@ -52,7 +52,7 @@ export function getForkedRepo(): string | undefined {
 	return select<HTMLMetaElement>('[name="octolytics-dimension-repository_parent_nwo"]')?.content;
 }
 
-export const parseTag = (tag: string): {version: string; namespace: string} => {
+export const parseTag = (tag: string): { version: string; namespace: string } => {
```
Comment (fregante) [inline source/github-helpers/index.ts:55]:
```suggestion
export const parseTag = (tag: string): {version: string; namespace: string} => {
```

File: source/github-helpers/index.ts [hunk 10]
```diff
@@ -107,6 +110,15 @@ export function preventPrCommitLinkLoss(url: string, pr: string, commit: string,
 	return `[\`${commit}\` (#${pr})](${url})`;
 }
 
+// To be  used as replacer callback in string.replace() for compare links
+export function preventPrCompareLinkLoss(url: string, compare: string, diff: string, index: number, fullText: string): string {
+	if (fullText[index + url.length] === ')') {
+		return url;
+	}
+
+	return `[\`${compare}\` ${diff}](${url})`;
```
Comment (fregante) [inline source/github-helpers/index.ts:119]:
```suggestion
	return `[\`${compare}\`${diff}](${url})`;
```

File: test/helpers.ts [hunk 11]
```diff
@@ -208,4 +214,27 @@ test('preventPrCommitLinkLoss', t => {
...
 		'It should not apply it twice even with hashes'
 	);
+	t.is(
+		replaceCompareLink('https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0'),
+		'https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0',
+		'It should not affect compare URLs without a diff hash'
+	);
+	t.is(
+		replaceCompareLink('https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191'),
+		'[`v11.5.2...v11.6.0` #diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191)'
```
Comment (fregante) [inline test/helpers.ts:224]:
```suggestion
		'[`v11.5.2...v11.6.0`#diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191)'
```

File: test/helpers.ts [hunk 12]
```diff
@@ -208,4 +214,27 @@ test('preventPrCommitLinkLoss', t => {
...
+	);
+	t.is(
+		replaceCompareLink('lorem ipsum dolor https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191 some random string'),
+		'lorem ipsum dolor [`v11.5.2...v11.6.0` #diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191) some random string'
+	);
+	t.is(
+		replaceCompareLink(replaceCompareLink('lorem ipsum dolor https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191 some random string')),
+		'lorem ipsum dolor [`v11.5.2...v11.6.0` #diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191) some random string',
```
Comment (fregante) [inline test/helpers.ts:232]:
```suggestion
		'lorem ipsum dolor [`v11.5.2...v11.6.0`#diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191) some random string',
```

File: test/helpers.ts [hunk 13]
```diff
@@ -208,4 +214,27 @@ test('preventPrCommitLinkLoss', t => {
...
+		'https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0',
+		'It should not affect compare URLs without a diff hash'
+	);
+	t.is(
+		replaceCompareLink('https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191'),
+		'[`v11.5.2...v11.6.0` #diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191)'
+	);
+	t.is(
+		replaceCompareLink('lorem ipsum dolor https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191 some random string'),
+		'lorem ipsum dolor [`v11.5.2...v11.6.0` #diff-6be2971b2b](https://github.com/sindresorhus/got/compare/v11.5.2...v11.6.0#diff-6be2971b2bb8dbf48d15ff680dd898b0R191) some random string'
```
Comment (fregante) [inline test/helpers.ts:228]:
```suggestion
		replaceCompareLink('lorem ipsum dolor https://github.com/sindresorhus/refined-github/compare/master...incremental-tag-changelog-link#diff-5b3cf6bcc7c5b1373313553dc6f93a5eR7-R9 some random string'),
		'lorem ipsum dolor [`master...incremental-tag-changelog-link`#diff-5b3cf6bcc7](https://github.com/sindresorhus/refined-github/compare/master...incremental-tag-changelog-link#diff-5b3cf6bcc7c5b1373313553dc6f93a5eR7-R9) some random string'
```

File: source/github-helpers/index.ts [hunk 14]
```diff
@@ -98,6 +98,9 @@ const escapeRegex = (string: string): string => string.replace(/[\\^$.*+?()[\]{}
 const prCommitPathnameRegex = /[/][^/]+[/][^/]+[/]pull[/](\d+)[/]commits[/]([\da-f]{7})[\da-f]{33}(?:#[\w-]+)?\b/; // eslint-disable-line unicorn/better-regex
 export const prCommitUrlRegex = new RegExp('\\b' + escapeRegex(location.origin) + prCommitPathnameRegex.source, 'gi');
 
+const prComparePathnameRegex = /[/][^/]+[/][^/]+[/]compare[/](.+)(#diff-[\da-f]{10})[\da-fR]{26}/; // eslint-disable-line unicorn/better-regex
```
Comment (fregante) [inline source/github-helpers/index.ts:101]:
```suggestion
const prComparePathnameRegex = /[/][^/]+[/][^/]+[/]compare[/](.+)(#diff-[\da-fR-]+)/; // eslint-disable-line unicorn/better-regex
```

File: source/github-helpers/index.ts [hunk 15]
```diff
@@ -107,6 +110,15 @@ export function preventPrCommitLinkLoss(url: string, pr: string, commit: string,
 	return `[\`${commit}\` (#${pr})](${url})`;
 }
 
+// To be  used as replacer callback in string.replace() for compare links
+export function preventPrCompareLinkLoss(url: string, compare: string, diff: string, index: number, fullText: string): string {
+	if (fullText[index + url.length] === ')') {
+		return url;
+	}
+
+	return `[\`${compare}\`${diff}](${url})`;
```
Comment (fregante) [inline source/github-helpers/index.ts:119]:
```suggestion
export function preventPrCompareLinkLoss(url: string, compare: string, hash: string, index: number, fullText: string): string {
	if (fullText[index + url.length] === ')') {
		return url;
	}

	return `[\`${compare}\`${hash.slice(0, 5)}](${url})`;
```

File: source.pem (hunk 7 above)
Comment (fregante) [reply source.pem]: I suggest you use [GitHub Desktop](https://desktop.github.com) or similar to commit only the files and lines that are necessary — and then look at the PR and see if there's any unwanted code. We shouldn't have to comment on this. This file is unrelated and its changes should be undone

File: test/helpers.ts [hunk 16]
```diff
@@ -11,7 +11,9 @@ import {
 	getScopedSelector,
 	getLatestVersionTag,
 	preventPrCommitLinkLoss,
-	prCommitUrlRegex
+	prCommitUrlRegex,
+	prCompareUrlRegex,
+	preventPrCompareLinkLoss
```
Comment (yakov116) [inline test/helpers.ts:16]:
```suggestion
```
This was imported 2 times

Comment (esakki-selvaraj) [reply test/helpers.ts:16]: It is done only once.

Comment (esakki-selvaraj) [reply test/helpers.ts:16]: I think you mistaken preventPrCommitLinkLoss ?

Comment (yakov116) [reply test/helpers.ts:16]: Your right!

Comment (fregante) [inline test/helpers.ts:16]: Let's move them together so nobody will confuse them

```suggestion
	preventPrCompareLinkLoss,
	prCommitUrlRegex,
	prCompareUrlRegex
```

**Review by fregante — APPROVED** [review]
Untested but looks fine

**Comment by fregante** [comment]
Tested and added proper test content and demo gif

**Comment by yakov116** [comment]
Does the link need to be updated (is goes to the old issue) maybe just update that issue?

**Comment by fregante** [comment]
> Does the link need to be updated (is goes to the old issue) maybe just update that issue?

Ideally the issue would properly describe the situation with gifs, demo broken content and what it does instead of being a bug report/feature request.

Feel free to create a new issue with that content if you feel like documenting it

**Comment by fregante** [comment]
# 🤝
